### ISSUE TITLE
Tebako press keeps failing with compilation error (MacOS, M2)

### ISSUE THREAD

**Issue by d4rky-pl** [tamatebako/tebako#161]
I'm trying to build Tebako on CI but it keeps failing and I'm no longer sure if it's me, CI, Tebako or dwarfs. The error:

```
/Users/semaphore/.tebako/deps/src/_dwarfs_wr/deps/src/_dwarfs/folly/folly/Conv.cpp:241:5: error: no template named 'bool_constant'
    bool_constant<'A' == 65 && 'Z' == 90 && 'a' == 97 && 'z' == 122>;
    ^
/Users/semaphore/.tebako/deps/src/_dwarfs_wr/deps/src/_dwarfs/folly/folly/Conv.cpp:246:10: error: use of undeclared identifier 'IsAscii'
  return IsAscii::value ? in | 0x20 : char(std::tolower(in));
         ^
-- Looking for 4 include files sys/types.h, ..., dlfcn.h - found
-- Looking for 5 include files sys/types.h, ..., errno.h
2 errors generated.
make[9]: *** [folly/CMakeFiles/folly_base.dir/folly/Conv.cpp.o] Error 1
make[8]: *** [folly/CMakeFiles/folly_base.dir/all] Error 2
make[8]: *** Waiting for unfinished jobs....
```

and then

```
[  7%] Completed '_libarchive'
[  7%] Built target _libarchive
make[7]: *** [all] Error 2
make[6]: *** [/Users/semaphore/.tebako/deps/src/_dwarfs_wr/deps/src/_dwarfs-stamp/_dwarfs-build] Error 2
make[5]: *** [CMakeFiles/_dwarfs.dir/all] Error 2
make[4]: *** [all] Error 2
make[3]: *** [/Users/semaphore/.tebako/deps/src/_dwarfs_wr-stamp/_dwarfs_wr-build] Error 2
make[2]: *** [CMakeFiles/_dwarfs_wr.dir/all] Error 2
make[1]: *** [CMakeFiles/tebako.dir/rule] Error 2
make: *** [tebako] Error 2
Tebako script failed: 'tebako press' build step failed [104]
```

The machine is Semaphore CI M2 Apple machine (`a2-standard-4`), I've installed all dependencies from README. Ruby version 3.2.4 but also fails on 3.2.2 (did not test other ones)

**Comment by ronaldtse** [comment]
Thank you @d4rky-pl for reporting this! We have not tried Semaphore CI so we can't rule out it's a platform issue but let us investigate.

It looks like a compiler compatibility issue.

@maxirmx could you please help? Thanks.

**Comment by maxirmx** [comment]
Thank you @d4rky-pl 
It happens in facebook code that is used to build dwarfs. 
I am running the tests against XCode 13 and 14 but it looks like Sempahore CI provedes XCode 15 in your configuration. Please give me some time, I will look at it.

**Comment by d4rky-pl** [comment]
@maxirmx thanks a lot for quick response 🙏  I just checked locally to confirm and it doesn't seem to be Semaphore CI specific and could be Xcode 15 related (possibly caused by clang 15, I was unable to build ruby-packer with that version too, though obviously with wildly different errors)

**Comment by maxirmx** [comment]
@d4rky-pl, I was not able to reproduce with XCode but I guess you are right, it looks like an issue with higher clang version
It may take a couple of days to fix

**Comment by ronaldtse** [comment]
@maxirmx is this an incompatibility of folly with clang-15?

**Comment by maxirmx** [comment]
This incompatibility of folly version we are using. 
I see that something is fixed exactly at the location where the issue occurs.

We will need to change folly version. It is painful,  because folly does not support musl and does not link in ridk environment and dwarfs does not support macos. So we have patch scripts that are universal in the large but require some tweaks and folly version upgrade implies full testing.

I plan to upgrade folly in a couple of days and I believe it will resolve the issue.

**Comment by maxirmx** [comment]
@d4rky-pl 

Could you please try ```gem install tebako --version 0.7.3.rc4``` ?

Although I fixed CI issue that you have pointed to I was not able to reproduce this issue on GitHub.  Installing XCode 15.4 locally did not help to reproduce it either. There is a factor that I do not understand but there is a chance that upgrade of facebook code provides a fix.
If not, please send me a full log from failed tebako run.  Also I will appreciate if you can share your Semaphore CI scrypt to the exted related to tebako. 

Thank you

**Comment by d4rky-pl** [comment]
@maxirmx the problem could be architecture related, your CI is running with Intel and I was testing on M2 Semaphore machine (`a2-standard-4`) that they let us test for a week. And if we're talking about [this libdwarfs build](https://github.com/tamatebako/libdwarfs/actions/runs/9978128760/job/27574321615), the build step took 25s so I think it may still be overcaching things and didn't actually try to compile anything.

With that said, I can confirm that the RC4 version does not fail on building `folly` anymore when running locally. It still fails to build but this is likely unrelated:

```
tebako-packager failed: Failed to run /Users/matyas/.tebako/o/s/bin/bundle exec /Users/matyas/.tebako/o/s/bin/gem build /Users/matyas/htdocs/helpling_toolbelt/helpling_toolbelt.gemspec (pid 79189 exit 1):
 /Users/matyas/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/definition.rb:540:in `materialize': Could not find jaro_winkler-1.5.6, ffi-1.15.5 in locally installed gems (Bundler::GemNotFound)
	from /Users/matyas/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/definition.rb:200:in `specs'
	from /Users/matyas/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/definition.rb:266:in `specs_for'
	from /Users/matyas/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/runtime.rb:18:in `setup'
	from /Users/matyas/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler.rb:162:in `setup'
	from /Users/matyas/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/setup.rb:23:in `block in <top (required)>'
	from /Users/matyas/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/ui/shell.rb:159:in `with_level'
	from /Users/matyas/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/ui/shell.rb:111:in `silence'
	from /Users/matyas/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/setup.rb:23:in `<top (required)>'
	from <internal:/Users/matyas/.tebako/o/s/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:86:in `require'
	from <internal:/Users/matyas/.tebako/o/s/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:86:in `require'
	from /Users/matyas/.tebako/o/s/lib/ruby/3.2.0/rubygems.rb:1370:in `<top (required)>'
	from <internal:gem_prelude>:2:in `require'
	from <internal:gem_prelude>:2:in `<internal:gem_prelude>'
 [255]
```

**Comment by maxirmx** [comment]
Thank you @d4rky-pl 

- Our CI runs on Intel and M1,  GHA macos-14 is an arm runner.  Also I have Intel and M1 on my desk :)
- Here is a build without cache https://github.com/tamatebako/libdwarfs/actions/runs/9978692318/job/27576127588 that consistently picks XCode 15.4 and clang 15
- Your log shows that tebako fails to deploy the solution to pristine environment. Should you need a packager we are interested to fix it.

**Comment by d4rky-pl** [comment]
@maxirmx I'm juggling production emergencies today (Rails upgrade 🥲) but I'll try to poke around more during the weekend. In the meantime, here's my build log after running `tebako press -R 3.2.4 -r $(pwd) -e exe/ht | tee buildlog.txt`. I would share the project as well but it's not currently open source and it honestly could be a problem with my local branch.

[buildlog.txt](https://github.com/user-attachments/files/16284015/buildlog.txt)

**Comment by d4rky-pl** [comment]
Side note: `export PATH="$(brew --prefix bison)/bin:$PATH"` should be added to README or the build process fails with 

```
ripper.y:12.10-14: require bison 3.0, but have 2.3
make[6]: *** [ripper.c] Error 63
make[5]: *** [ext/ripper/static] Error 2
```

because homebrew for bison is keg-only. I keep tripping over this every time I'm retesting things in a fresh environment 😅

**Comment by d4rky-pl** [comment]
@maxirmx I thought it might be an issue with my local environment (for ex. clashing with asdf) but it seems that `tebako press` fails also on CI with the same error. Should I report it as a separate issue?

```
   ... @ /Users/semaphore/.tebako/o/s/bin/bundle install --jobs=4
   ... @ /Users/semaphore/.tebako/o/s/bin/bundle exec /Users/semaphore/.tebako/o/s/bin/gem build /Users/semaphore/helpling_toolbelt/release/helpling_toolbelt.gemspec
tebako-packager failed: Failed to run /Users/semaphore/.tebako/o/s/bin/bundle exec /Users/semaphore/.tebako/o/s/bin/gem build /Users/semaphore/helpling_toolbelt/release/helpling_toolbelt.gemspec (pid 42347 exit 1):
 /Users/semaphore/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/definition.rb:540:in `materialize': Could not find jaro_winkler-1.5.6, ffi-1.15.5 in locally installed gems (Bundler::GemNotFound)
	from /Users/semaphore/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/definition.rb:200:in `specs'
	from /Users/semaphore/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/definition.rb:266:in `specs_for'
	from /Users/semaphore/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/runtime.rb:18:in `setup'
	from /Users/semaphore/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler.rb:162:in `setup'
	from /Users/semaphore/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/setup.rb:23:in `block in <top (required)>'
	from /Users/semaphore/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/ui/shell.rb:159:in `with_level'
	from /Users/semaphore/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/ui/shell.rb:111:in `silence'
	from /Users/semaphore/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.4.18/lib/bundler/setup.rb:23:in `<top (required)>'
	from <internal:/Users/semaphore/.tebako/o/s/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:86:in `require'
	from <internal:/Users/semaphore/.tebako/o/s/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:86:in `require'
	from /Users/semaphore/.tebako/o/s/lib/ruby/3.2.0/rubygems.rb:1370:in `<top (required)>'
	from <internal:gem_prelude>:2:in `require'
	from <internal:gem_prelude>:2:in `<internal:gem_prelude>'
 [255]
make[3]: *** [CMakeFiles/source_filesystem] Error 255
make[2]: *** [CMakeFiles/source_filesystem.dir/all] Error 2
make[1]: *** [CMakeFiles/tebako.dir/rule] Error 2
make: *** [tebako] Error 2
Tebako script failed: 'tebako press' build step failed [104]
```

**Comment by d4rky-pl** [comment]
@maxirmx reported #164 as a separate issue. The rc4 version of the gem works for me otherwise so feel free to close this bug after the release ❤️

**Comment by maxirmx** [comment]
This pease of code looks like an error for me
```
void Executor::invokeCatchingExnsLog(char const* const prefix) noexcept {
  auto ep = current_exception();
  LOG(ERROR) << prefix << " threw unhandled " << exceptionStr(ep);
}
```

It shall be ```std::current_exception()``` since ```using namespace std``` is missing

I guess there is some legacy disposition of includes where ```using namespace std``` is present in one of include files but at Semaphore CI you have newer version. Also it is possible that some (body of) compiler versions correctly treat ```current_exception``` without ```std::```

You may be right that behaviour has changed on M2.

I can patch folly on the fly, anyway I am doing it for a couple of other issues.  It will take a couple of days

**Comment by d4rky-pl** [comment]
@maxirmx I just realized I was running tebako 0.7.4 on CI and 0.8.0 locally, after re-running with 0.8.0 it seems to work now 🤦  Either that or your patch already worked, in any case I finally managed to build it on CI now

**Comment by maxirmx** [comment]
With 0.8.0 you get newer version of folly

**Comment by d4rky-pl** [comment]
@maxirmx I finally figured it out so I'll leave the answer if someone searches for it in the future:

On CI for the optimization sake we had these two ENV vars enabled:

```sh
export HOMEBREW_NO_INSTALL_CLEANUP=1
export HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK=1
```

The latter caused Homebrew to not upgrade dependent packages when installing dependencies (that was the difference between my local machine, CI and CI when debugging via ssh and running commands one by one). Removing these ENV vars made the builds pass but also caused them to be much slower as Homebrew by default tries to upgrade everything, including completely unnecessary things (openjdk, awscli, two versions of python etc.) so I checked the packages case by case instead.

It seems that the one preventing the build from completing was... system `folly`. After adding it to the list of brew dependencies and letting it upgrade from `2024.04.08.00` to `2024.07.15.00` the problem has resolved itself and things started building reliably. Not sure if it makes sense to update README and include it there as this is a bit of an optimization edge case (brew would normally install it on its own) and I honestly have zero idea how this fixed the problem, only that it did.

Once again thanks a lot for your help and the continued maintenance of the project @maxirmx :pray:
